Set up metrics code (#45 & #46)

lmorchard opened the pull request:
Throwing this out there to start getting some eyes on it. I tried cribbing `docs/metrics.md` and the ping format [from fellow webextension-style experiment blok](https://github.com/mozilla/blok/blob/master/docs/metrics.md), but it could use some review & further tweaking.

Will poke some other Test Pilot folks next week for opinons - off the top of my head, @johngruen & @6a68 & @clouserw might have some?

lmorchard commented:
Also, the tests here don't actually test `background.js` - but as I tried getting into that, it seemed like I'd have to restructure a bunch of it into a submodule that could get mock `browser` APIs injected via `init()` and I figured that's too ocean-boiling-esque for this PR.

wresuolc commented on docs/metrics.md:
```diff
@@ -0,0 +1,113 @@
+There is currently no server side component to Snooze Tabs.
+
+### Client Side
+Snooze Tabs will use Test Pilot's Telemetry wrapper with no batching of data.
```
We'll modify this to say something about @_6a68 's broker once we have more details on it

wresuolc commented on docs/metrics.md:
````diff
@@ -0,0 +1,113 @@
+}
+```
+
+* A previously snoozed tab is snoozed again after waking
````
This implies that once a tab is snoozed you're forever tracking that fact about that tab?

lmorchard replied:
Hmm, yeah, maybe we should be more specific that it's only tracked the first time it's resnoozed. Or, until the browser restarts - which is what the code actually does right now, because I don't persist the tracking anywhere permanent.

lmorchard commented on docs/metrics.md:
````diff
@@ -0,0 +1,113 @@
+}
+```
+
+A Redshift schema for the payload:
````
Also, this schema is a total guess, since I've not done one before and haven't seen where it plugs in

wresuolc replied:
Seems nice and simple.  r? @6a68 and @sunahsuh

bwinton commented on src/background.js:
```diff
@@ -83,14 +85,23 @@ const idForItem = item => `${item.time}-${item.url}`;
+  },
   update: message =>
-    messageOps.cancel(message.old).then(() => messageOps.schedule(message.updated))
+    messageOps.cancel(message.old).then(() => messageOps.schedule(message.updated)),
```
Should this get a separate Metrics ping, or did we want to force people to back-calculate which cancel/schedule pairs were updates and which weren't?

lmorchard replied:
Not sure - updates weren't listed as one of the measurements in [the plan so far](https://github.com/bwinton/SnoozeTabs/issues/45). Seems like a handy thing to measure, though

bwinton replied:
Well, I would say let's do it, then, if we can.

bwinton commented on src/lib/components/MainPanel.js:
```diff
@@ -68,6 +68,6 @@ export default class MainPanel extends React.Component {
     const { scheduleSnoozedTab } = this.props;
     if (!dateChoice) { return; }
-    scheduleSnoozedTab(dateChoice);
+    scheduleSnoozedTab(dateChoice, 'pick');
```
I think we should use the constant from `lib/times.js` here instead.  (And add one for `'week'` down below.)

lmorchard replied:
Yeah, can revise for that. There wasn't a constant yet when I wrote this bit

bwinton commented on src/background.js:
```diff
@@ -33,6 +34,7 @@ function updateButtonForTab(tabId, changeInfo) {
 
 function init() {
   log('init()');
+  Metrics.init(window.BroadcastChannel, browser.tabs);
```
So, I'm not really sure why we're passing in the BroadcastChannel here.  I suspect it's the right thing to do, but could you elaborate a little more on it (maybe in a comment in `metrics.js`)?

lmorchard replied:
Basically, it's dependency injection, so that the tests don't have to rely on a `window` global to pass in a `BroadcastChannel` class because it's not an API based on an importable module

bwinton replied:
👍

bwinton commented on src/lib/metrics.js:
```diff
@@ -0,0 +1,99 @@
+let pingChannel = null;
+
+// Name of the metrics BroadcastChannel expected by the Test Pilot add-on
+const TESTPILOT_TELEMETRY_CHANNEL = 'testpilot-telemetry';
```
There seems to be a lot of Test-Pilot-specific _and_ SnoozeTabs-specific stuff in here.  How easy would it be to switch this over to use Google Analytics if it ends up graduating into a Shield study?  Would it make sense to split the Test-Pilot-specific stuff out into a separate file that the SnoozeTabs-specific stuff could call, and then switch that file out for a Google-Analytics-specific implementation?

lmorchard replied:
What I've got here now is just basically copying what the Tracking Protection experiment does now. But for Telemetry vs GA (or both), these are all things that will hopefully be addressed with [the pending metrics broker solution from Test Pilot](https://github.com/mozilla/testpilot/issues/1735). Not sure what that will look like yet, though

bwinton replied:
Sounds reasonable.

bwinton commented on test/.setup.js:
```diff
@@ -0,0 +1,24 @@
+global.navigator = {
+  userAgent: 'node.js'
+};
+*/
```
Do we need this file yet?  Should we not add it until we do?  🙂

lmorchard replied:
We'll need it the first time any React components are tested, and it's handy for setting up globals like `window` when they can't be avoided by doing a dependency injection thing

lmorchard commented:
Updated to remove `test/.setup.js` and to use the `PICK_TIME` constant in relevant spots.

lmorchard commented on docs/metrics.md:
```diff
@@ -0,0 +1,113 @@
+
+### Immediate Questions
+
+TBD
```
Also, need to formulate the questions here. Hoping we have them to be copypasta'd from a product doc somewhere. Or, I can gin them up myself if necessary because I know the best words

wresuolc replied:
This is from the Product Vision:
> How will we know we are successful?  Key metric here will be retained users. At what rate do experimenters snooze their tabs? Do users snooze tabs with some consistency or does the novelty wear of quickly?
but they could probably use some word tweaking.  /cc @johngruen @mattgrimes @sharonbautista

johngruen replied:
The basic question we seek to answer is: does snooze tabs provide a useful tool for its users in Test Pilot.

* Do people snooze tabs and at what rate? 
	* x = time, y = number of snoozes (vanity metric)
	* x = time, y = snoozes per user
* Do people close returned tabs without focusing them and at what rate?
	* x= time, y = snoozes focused and snoozes dismissed w/o focus
* Are the measurably different rates of returned tab engagement for snoozes of different lengths.
	* bar graph showing rate at which different snooze length tabs are focused and closed (should show custom times as well (these can be clustered into one group)
* What is the overall distribution of snooze times across all users?
	* Bar graph showing distribution of snooze times
* Do people set custom times, and at what rate, 
	* x = time, y = number of custom snooze events
* Do people re-snooze previously snoozed tabs?
	* x = time, y = snoozes of previously snoozed tabs per user

bwinton commented on package.json:
```diff
@@ -49,25 +53,27 @@
   "scripts": {
-    "start": "npm run build && npm-run-all --parallel watch run storybook",
+    "start": "npm run build && npm-run-all --parallel test watch run storybook",
     "start-win": "npm run build && npm run watch",
```
Let's add the test to `start-win`, too… (and change it to `npm-run-all --parallel` while we're at it).

lmorchard commented:
Rebased against master and added another ping for [when a snoozed tab is updated](https://github.com/bwinton/SnoozeTabs/pull/95/files#diff-683a405c5c005542f9e848b827fe0765R92).

lmorchard commented:
Considering merging this as-is, since it's big & bit-rotty but does actually work. Pending updates are on changes to semi-unrelated external things that we can roll in as separate issues

bwinton commented:
Given the rot, I'm 👍 on merging it and filing followup bugs for the other stuff.  (Let me know when the conflict is fixed, and I'll do the merge if you'ld like. 🙂)

wresuolc commented on docs/metrics.md:
```diff
@@ -0,0 +1,122 @@
+
+TBD
+
+### Follow-up Questions
```
we can drop this section

lmorchard commented:
Rebased, updated `docs/metrics.md`. Think I'll merge this and then file an issue to switch over to [testpilot-metrics](https://github.com/mozilla/testpilot-metrics) ASAP

lmorchard commented:
Now that this is merged, anyone listening should feel free to submit issues / PRs to further refine `docs/metrics.md`
